File: HPBarKit/Sources/HPBarKit/Services/Pricing.swift
import Foundation

/// Per-million-token price for one model, in USD.
///
/// `cacheCreate` covers the default cache-write rate (Anthropic's 5-minute TTL,
/// 1.25× input). `cacheCreate1h` is optional — when set, it applies to the
/// 1-hour TTL writes (2× input for Anthropic models). Non-Claude models that
/// don't have a TTL distinction can omit `cache_create_1h` and the 5m rate
/// will be used for both.
public struct ModelPrice: Codable, Sendable, Equatable {
    public let input: Double
    public let output: Double
    public let cacheRead: Double
    public let cacheCreate: Double
    public let cacheCreate1h: Double?

    public init(
        input: Double,
        output: Double,
        cacheRead: Double,
        cacheCreate: Double,
        cacheCreate1h: Double? = nil
    ) {
        self.input = input
        self.output = output
        self.cacheRead = cacheRead
        self.cacheCreate = cacheCreate
        self.cacheCreate1h = cacheCreate1h
    }

    enum CodingKeys: String, CodingKey {
        case input, output
        case cacheRead = "cache_read"
        case cacheCreate = "cache_create"
        case cacheCreate1h = "cache_create_1h"
    }
}

/// Per-component dollar cost for a `ModelUsage` row.
public struct ModelCost: Sendable, Equatable {
    public let input: Double
    public let output: Double
    public let cacheRead: Double
    public let cacheCreate: Double  // combined 5m + 1h
    public var total: Double { input + output + cacheRead + cacheCreate }
    public init(input: Double, output: Double, cacheRead: Double, cacheCreate: Double) {
        self.input = input; self.output = output
        self.cacheRead = cacheRead; self.cacheCreate = cacheCreate
    }
}

/// Table of model-id → ModelPrice. Built-in entries cover current Anthropic
/// models; users can override or extend (e.g. `kimi-k2.6`) by writing JSON to
/// `~/Library/Application Support/HPBar/pricing.json`:
///
/// ```json
/// {
///   "kimi-k2.6": {
///     "input": 0.50, "output": 2.00,
///     "cache_read": 0.10, "cache_create": 0.60,
///     "cache_create_1h": 0.95   // optional; omit if vendor has no TTL split
///   }
/// }
/// ```
///
/// All values are dollars per million tokens. Anything in the user file
/// overrides the built-in entry of the same id.
public struct Pricing: Sendable {
    public let table: [String: ModelPrice]

    public init(table: [String: ModelPrice]) { self.table = table }

    public func price(for modelId: String) -> ModelPrice? {
        // Fall back to a date-suffix-stripped id so `claude-haiku-4-5-20251001`
        // matches a `claude-haiku-4-5` entry.
        table[modelId] ?? table[Self.stripDateSuffix(modelId)]
    }

    /// Dollar cost split by component. Cache-write tokens are passed as two
    /// buckets because Anthropic charges 5m vs 1h cache writes at different
    /// rates (1.25× vs 2× the base input price).
    public func cost(
        for modelId: String,
        input: Int,
        output: Int,
        cacheRead: Int,
        cacheCreate5m: Int,
        cacheCreate1h: Int
    ) -> ModelCost? {
        guard let p = price(for: modelId) else { return nil }
        func c(_ tokens: Int, _ rate: Double) -> Double {
            Double(tokens) * rate / 1_000_000
        }
        let oneHourRate = p.cacheCreate1h ?? p.cacheCreate
        return ModelCost(
            input:       c(input,         p.input),
            output:      c(output,        p.output),
            cacheRead:   c(cacheRead,     p.cacheRead),
            cacheCreate: c(cacheCreate5m, p.cacheCreate) + c(cacheCreate1h, oneHourRate)
        )
    }

    // MARK: - Loading
    //
    // All prices live in JSON, not Swift. Two sources, merged:
    //   1. Bundled `pricing.json` — defaults shipped with the app
    //      (HPBarKit/Sources/HPBarKit/Resources/pricing.json). Source of
    //      truth for Anthropic models. Updated by app releases.
    //   2. `~/Library/Application Support/HPBar/pricing.json` — user
    //      overrides. Anything here wins over the bundled file. Use it to
    //      add custom models or correct rates without recompiling.

    /// Load the bundled table, then overlay user overrides. Missing or
    /// malformed files are tolerated (bundled missing → empty; user missing
    /// → just bundled).
    public static func loaded() -> Pricing {
        var merged = bundledTable()
        for (k, v) in userOverrides() { merged[k] = v }
        return Pricing(table: merged)
    }

    /// Empty table — useful for tests / DI.
    public static let empty = Pricing(table: [:])

    private static func bundledTable() -> [String: ModelPrice] {
        guard let url = Bundle.module.url(forResource: "pricing", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let table = try? JSONDecoder().decode([String: ModelPrice].self, from: data)
        else { return [:] }
        return table
    }

    private static func userOverrides() -> [String: ModelPrice] {
        guard let url = overrideURL(),
              let data = try? Data(contentsOf: url),
              let table = try? JSONDecoder().decode([String: ModelPrice].self, from: data)
        else { return [:] }
        return table
    }

    private static func stripDateSuffix(_ id: String) -> String {
        var parts = id.split(separator: "-").map(String.init)
        while let last = parts.last, last.count >= 6, last.allSatisfy(\.isNumber) {
            parts.removeLast()
        }
        return parts.joined(separator: "-")
    }

    private static func overrideURL() -> URL? {
        let fm = FileManager.default
        guard let supp = try? fm.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: false
        ) else { return nil }
        return supp.appendingPathComponent("HPBar/pricing.json")
    }
}

File: tauri/src-tauri/src/pricing.rs
//! Per-model token pricing. Port of the Swift `Pricing` / `ModelPrice`.
//!
//! Prices live in JSON, not code. Two sources, merged:
//!   1. Bundled `pricing.json` (compiled in via `include_str!`) — defaults.
//!   2. User overrides at the platform data dir (e.g. macOS
//!      `~/Library/Application Support/HPBar/pricing.json`). Anything here wins.
//!
//! All values are dollars per million tokens.

use serde::Deserialize;
use std::collections::HashMap;

const BUNDLED: &str = include_str!("pricing.json");

#[derive(Deserialize, Clone, Debug)]
pub struct ModelPrice {
    pub input: f64,
    pub output: f64,
    pub cache_read: f64,
    pub cache_create: f64,
    /// 1-hour-TTL cache writes (2× input for Anthropic). Falls back to
    /// `cache_create` when absent.
    pub cache_create_1h: Option<f64>,
}

/// Per-component dollar cost for a model's usage.
#[derive(Clone, Debug, Default)]
pub struct ModelCost {
    pub input: f64,
    pub output: f64,
    pub cache_read: f64,
    pub cache_create: f64, // combined 5m + 1h
}

impl ModelCost {
    pub fn total(&self) -> f64 {
        self.input + self.output + self.cache_read + self.cache_create
    }
}

pub struct Pricing {
    table: HashMap<String, ModelPrice>,
}

impl Pricing {
    /// Bundled table overlaid with user overrides. Malformed/missing files are
    /// tolerated (bundled wins nothing → empty; user missing → just bundled).
    pub fn loaded() -> Self {
        let mut table: HashMap<String, ModelPrice> =
            serde_json::from_str(BUNDLED).unwrap_or_default();
        for (k, v) in user_overrides() {
            table.insert(k, v);
        }
        Pricing { table }
    }

    fn price(&self, model_id: &str) -> Option<&ModelPrice> {
        self.table
            .get(model_id)
            .or_else(|| self.table.get(&strip_date_suffix(model_id)))
    }

    /// Dollar cost split by component. Cache writes come as two buckets because
    /// 5m and 1h writes bill at different rates.
    pub fn cost(
        &self,
        model_id: &str,
        input: i64,
        output: i64,
        cache_read: i64,
        cache_create_5m: i64,
        cache_create_1h: i64,
    ) -> Option<ModelCost> {
        let p = self.price(model_id)?;
        let c = |tokens: i64, rate: f64| (tokens as f64) * rate / 1_000_000.0;
        let one_hour_rate = p.cache_create_1h.unwrap_or(p.cache_create);
        Some(ModelCost {
            input: c(input, p.input),
            output: c(output, p.output),
            cache_read: c(cache_read, p.cache_read),
            cache_create: c(cache_create_5m, p.cache_create) + c(cache_create_1h, one_hour_rate),
        })
    }
}

fn user_overrides() -> HashMap<String, ModelPrice> {
    let Some(dir) = dirs::data_dir() else {
        return HashMap::new();
    };
    let path = dir.join("HPBar").join("pricing.json");
    std::fs::read_to_string(path)
        .ok()
        .and_then(|s| serde_json::from_str(&s).ok())
        .unwrap_or_default()
}

/// `claude-haiku-4-5-20251001` → `claude-haiku-4-5` (drop trailing date stamp).
pub fn strip_date_suffix(id: &str) -> String {
    let mut parts: Vec<&str> = id.split('-').collect();
    while let Some(last) = parts.last() {
        if last.len() >= 6 && last.chars().all(|c| c.is_ascii_digit()) {
            parts.pop();
        } else {
            break;
        }
    }
    parts.join("-")
}
